pkg/urbit/vere/io/conn.c

```diff
@@ -0,0 +1,918 @@
+*/
+static void
+_conn_ovum_bail(u3_ovum* egg_u, u3_noun lud)
+{
+  u3_cran* ran_u = egg_u->ptr_v;
+  u3_chan* can_u = ran_u->can_u;
+
+  u3_auto_bail_slog(egg_u, u3k(lud));
+  if ( !can_u ) {
+    //  chan was closed; noop.
+    //
+    u3z(ran_u->rid); c3_free(ran_u);
+    u3z(lud); return;
+  }
+  _conn_send_noun(can_u, u3nt(ran_u->rid, c3__bail, lud));
+  _conn_drop_cran(can_u, ran_u);
+  u3_ovum_free(egg_u);
+}
+
+/* _conn_ovum_news(): lifecycle callback for injected events.
+*/
+static void
+_conn_ovum_news(u3_ovum* egg_u, u3_ovum_news new_e)
+{
+  u3_cran*  ran_u = egg_u->ptr_v;
+  u3_chan*  can_u = ran_u->can_u;
+
+  if ( u3_ovum_done == new_e ||
+       u3_ovum_drop == new_e )
+  {
+    if ( can_u ) {
+      _conn_send_noun(can_u,
+                      u3nt(ran_u->rid, c3__news,
+                           ( u3_ovum_done == new_e
+                             ? c3__done
+                             : c3__drop )));
+      _conn_drop_cran(can_u, ran_u);
+    }
+    else {
+      u3z(ran_u->rid); c3_free(ran_u);
+    }
+  }
+}
+
+/* _conn_read_peel(): response to a %peel request, sans rid.
+*/
+static u3_noun
+_conn_read_peel(u3_conn* con_u, u3_noun dat)
+{
+  u3_pier*  pir_u = con_u->car_u.pir_u;
+  u3_noun   i_dat, t_dat, it_dat, tt_dat;
+  u3_noun   res;
+
+  if ( c3n == u3r_cell(dat, &i_dat, &t_dat) ) {
+    res = u3_nul;
+  }
+  else if ( u3_nul == t_dat ) {
+    //  zero-argument requests.
+    //
+    switch (i_dat) {
+      default: {
+        res = u3_nul;
+      } break;
+      //  command list.
+      //
+      case c3__help: {
+        res = u3nc(
+          u3_nul,
+          u3i_list(
+            u3nc(c3__help, u3_nul),
+            u3nc(c3__khan, u3_nul),
+            u3nc(c3__live, u3_nul),
+            u3nc(c3__mass, u3_nul),
+            u3nc(c3__port, u3nq(c3__ames, c3__htls, c3__http, u3_nul)),
+            u3nc(c3__v, u3_nul),
+            u3nc(c3__vars, u3_nul),
+            u3_none));
+      } break;
+      //  simple health check.
+      //
+      case c3__live: {
+        res = u3nc(u3_nul, pir_u->liv_o);
+      } break;
+      //  true iff the %khan vane is live (meaning %fyrd is supported.)
+      //
+      case c3__khan: {
+        res = u3nc(u3_nul, con_u->kan_o);
+      } break;
+      //  |mass output
+      //
+      case c3__mass: {
+        //  TODO  |mass
+        //
+        res = u3_nul;
+      } break;
+      //  runtime metrics list.
+      //
+      case c3__vars: {
+        u3_auto* car_u;
+
+        res = u3_nul;
+        if ( u3_psat_work != pir_u->sat_e ) {
+          //  empty response if drivers are not live.
+          //
+          break;
+        }
+        car_u = pir_u->wok_u->car_u;
+        while ( car_u ) {
+          if ( car_u->io.info_f ) {
+            res = u3nc(
+              u3nc(car_u->nam_m, car_u->io.info_f(car_u)),
+              res);
+          }
+          car_u = car_u->nex_u;
+        }
+        res = u3nc(u3_nul, res);
+      } break;
```

Comment:
I think we want to move this into `u3_auto_info()`, and call that from a `u3_pier_info()`, which should do the check for `u3_psat_work`.

Reply:
done. `u3_pier_slog` (previously named `…_info`) does some extra work, which I mostly replicated in the new `u3_pier_info`. i skipped disk and lord info for now for expedience and since we may still change the output type.